### PR TITLE
Generate: fix logits processors doctests

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

The doctests got stale 👀 (related PR to prevent this from happening again: https://github.com/huggingface/transformers/pull/29716)

There are 2 main categories of fixes:
1. Fixes where there is randomness involved: updates seed and potentially the output of the "bad" example. I can reproduce the existing doctest results if I go back to an older version (like `v4.35`), but I don't think it's worth diving through to find the root cause, as many harmless things can change the output of sampling;
2. Whisper fixes (cc @sanchit-gandhi )

All tests are passing after these changes (`pytest --doctest-modules src/transformers/generation/logits_process.py -vv`)

### CODE REVIEW

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -1204,16 +1210,16 @@ class PrefixConstrainedLogitsProcessor(LogitsProcessor):
...
-    >>> entity =  tokenizer(" Bob Marley", return_tensors="pt").input_ids[0]  # 3 tokens
+    >>> entity = tokenizer(" Bob Marley", return_tensors="pt").input_ids[0]  # 3 tokens
     >>> def prefix_allowed_tokens_fn(batch_id, input_ids):
     ...     '''
     ...     Attempts to generate 'Bob Marley' when 'Bob' is detected.
     ...     In this case, `batch_id` is not used, but you can set rules for each batch member.
     ...     '''
     ...     if input_ids[-1] == entity[0]:
-    ...         return entity[1]
+    ...         return [entity[1].item()]
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1220]: The signature of `prefix_allowed_tokens_fn` should be `Callable[[int, torch.Tensor], List[int]]`, as explained in the docs

File: src/transformers/models/whisper/generation_whisper.py [hunk 2]
```diff
@@ -474,11 +473,8 @@ def generate(
                 "The input name `inputs` is deprecated. Please make sure to use `input_features` instead.",
                 FutureWarning,
             )
-        # 1. copy generation config
-        if generation_config is None:
-            generation_config = copy.deepcopy(self.generation_config)
-        else:
-            generation_config = copy.deepcopy(generation_config)
+        # 1. prepare generation config
+        generation_config, kwargs = self._prepare_generation_config(generation_config, **kwargs)
```
Comment (amyeroberts) [inline src/transformers/models/whisper/generation_whisper.py:477]: The lines above imply there's a `self.generation_config` which should be used if `generation_config` is `None`

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -414,7 +414,7 @@ class TopPLogitsWarper(LogitsWarper):
     ```python
     >>> from transformers import AutoTokenizer, AutoModelForCausalLM, set_seed
 
-    >>> set_seed(0)
+    >>> set_seed(1)
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:417]: Why change the seed?

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -1641,7 +1647,7 @@ class SuppressTokensAtBeginLogitsProcessor(LogitsProcessor):
     >>> # Whisper has `begin_suppress_tokens` set by default (= `[220, 50256]`). 50256 is the EOS token, so this means
     >>> # it can't generate and EOS token in the first iteration, but it can in the others.
     >>> outputs = model.generate(**inputs, return_dict_in_generate=True, output_scores=True)
-    >>> print(outputs.scores[1][0, 50256])  # 1 (and not 0) is the first freely generated token
+    >>> print(outputs.scores[0][0, 50256])
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1650]: Whisper changes: @sanchit-gandhi let me know if they make sense, according to recent changes in Whisper

File: src/transformers/generation/logits_process.py [hunk 5]
```diff
@@ -478,7 +480,7 @@ class TopKLogitsWarper(LogitsWarper):
     ```python
     >>> from transformers import AutoTokenizer, AutoModelForCausalLM, set_seed
 
-    >>> set_seed(0)
+    >>> set_seed(1)
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:483]: Same here

File: src/transformers/generation/logits_process.py [hunk 6]
```diff
@@ -1714,36 +1720,6 @@ class ForceTokensLogitsProcessor(LogitsProcessor):
     indices that will be forced before generation. The processor will set their log probs to `inf` so that they are
     sampled at their corresponding index. Originally created for
     [Whisper](https://huggingface.co/docs/transformers/model_doc/whisper).
-
-    Examples:
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1718]: This processor is going to be removed in `v4.40`, so I didn't want to spend time fixing the test :D

File: src/transformers/generation/logits_process.py [hunk 7]
```diff
@@ -619,7 +621,7 @@ class EpsilonLogitsWarper(LogitsWarper):
     ```python
     >>> from transformers import AutoTokenizer, AutoModelForCausalLM, set_seed
 
-    >>> set_seed(0)
+    >>> set_seed(1)
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:624]: Same comment about the seed

File: src/transformers/generation/logits_process.py [hunk 8]
```diff
@@ -696,7 +700,7 @@ class EtaLogitsWarper(LogitsWarper):
...
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:703]: seed change here too

File: src/transformers/models/whisper/generation_whisper.py (hunk 2 above)
Comment (gante) [inline src/transformers/models/whisper/generation_whisper.py:477]: This function from the main `generate` body (`_prepare_generation_config`) pulls generation parameterization from `kwargs` into `generation_config`. 

Some Whisper-based doctests were incorrect without this functionality.

File: src/transformers/generation/logits_process.py [hunk 9]
```diff
@@ -423,7 +423,9 @@ class TopPLogitsWarper(LogitsWarper):
     >>> # With sampling, the output is unexpected -- sometimes too unexpected.
     >>> outputs = model.generate(**inputs, do_sample=True)
     >>> print(tokenizer.batch_decode(outputs, skip_special_tokens=True)[0])
-    A sequence: 1, 2, 0, 2, 2. 2, 2, 2, 2
+    A sequence: 1, 2, 3 | < 4 (left-hand pointer) ;
+    <BLANKLINE>
+    <BLANKLINE>
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:428]: Do we expect the generations to be so much worse (even in the sampling case)? It seems strange that they're so different

File: src/transformers/generation/logits_process.py [hunk 10]
```diff
@@ -1714,36 +1720,6 @@ class ForceTokensLogitsProcessor(LogitsProcessor):
...
-
-    >>> # If we disable `forced_decoder_ids`, we stop seeing that effect
-    >>> outputs = model.generate(**inputs, return_dict_in_generate=True, output_scores=True, forced_decoder_ids=None)
-    >>> print(
-    ...     all(outputs.scores[0][0, i] == float("-inf") for i in range(processor.tokenizer.vocab_size) if i != 50362)
-    ... )
-    False
-    >>> print(outputs.scores[0][0, 50362])
-    tensor(19.3140)
-    ```
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:1746]: Why remove the example here?

File: src/transformers/generation/logits_process.py [hunk 11]
```diff
@@ -1604,13 +1610,13 @@ class LogitNormalization(LogitsProcessor, LogitsWarper):
     >>> # By default, the scores are not normalized -- the sum of their exponentials is NOT a normalized probability
     >>> # distribution, summing to 1
     >>> outputs = model.generate(**inputs, return_dict_in_generate=True, output_scores=True)
-    >>> print(torch.sum(torch.exp(outputs.scores[-1])))
-    tensor(816.3250)
+    >>> print(torch.allclose(torch.sum(torch.exp(outputs.scores[-1])), torch.Tensor((1.000,)), rtol=1e-4))
+    False
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:1614]: The previous output was more informative imo - there's infinitely many ways to not be close to 1

File: src/transformers/generation/logits_process.py [hunk 12]
```diff
@@ -1641,7 +1647,7 @@ class SuppressTokensAtBeginLogitsProcessor(LogitsProcessor):
     >>> # Whisper has `begin_suppress_tokens` set by default (= `[220, 50256]`). 50256 is the EOS token, so this means
     >>> # it can't generate and EOS token in the first iteration, but it can in the others.
     >>> outputs = model.generate(**inputs, return_dict_in_generate=True, output_scores=True)
-    >>> print(outputs.scores[1][0, 50256])  # 1 (and not 0) is the first freely generated token
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:1644]: out of interest - what changed here?

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this fix! 

I have a few questions about the changes, in particular why we need to change the seed

File: src/transformers/models/whisper/generation_whisper.py (hunk 2 above)
Comment (gante) [reply src/transformers/models/whisper/generation_whisper.py:477]: `_prepare_generation_config` does precisely that: https://github.com/huggingface/transformers/blob/484e10f7f26b62452ba96488811a0c3f6eaf58df/src/transformers/generation/utils.py#L1204

It is a more complex version of this original if/else that preserves additional backward (and forward!) compatibility features of `generate` :)

File: src/transformers/generation/logits_process.py (hunk 3 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:417]: The seed is changed because the sample output is changed (more on that below), and a new seed was selected to illustrate the point of the example 🤗 I wanted a seed that produced a bad output in the unparameterized call and a good output in the parameterized call. Bear in mind that the model used in the examples is very small, and thus noisy with sampling.

We need to change the seed because the output of sampling has changed. There are many innocuous changes that can cause this: tiny numerical differences due to different versions, tiny numerical differences due to reordering of operations, corrections in the architecture code, different RNG behavior in torch (unlikely), and so on. As I've written in the PR header, I don't think it's worth our time finding the exact cause. The results in most other sampling tests are unchanged, there are many innocuous changes that can cause this, and it may be time-consuming to pin the cause.

File: src/transformers/generation/logits_process.py (hunk 9 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:428]: Yes, the model is very small :)

File: src/transformers/generation/logits_process.py (hunk 10 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:1746]: > This processor is going to be removed in v4.40, so I didn't want to spend time fixing the test :D

:)

File: src/transformers/generation/logits_process.py (hunk 11 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:1614]: True, but it is beyond the scope of the example -- the key point here is adding the flag normalizes the probability distribution.

Testing against the exact number caused the test to fail. In fact, if we run this test on different hardware (local compute vs DGX), we get a slightly different number. We could work around it with `torch.allclose`, but I don't think it adds value to the test :)

File: src/transformers/generation/logits_process.py (hunk 12 above)
Comment (gante) [reply src/transformers/generation/logits_process.py:1644]: I believe the indexing of first freely decoded token changed recently in Whisper, but I'd like to have @sanchit-gandhi confirming the correctness of these changes :)

Comment (amyeroberts) [reply src/transformers/generation/logits_process.py:1644]: This might be a possible BC issue :/

File: src/transformers/generation/logits_process.py (hunk 11 above)
Comment (amyeroberts) [reply src/transformers/generation/logits_process.py:1614]: OK 👍

File: src/transformers/generation/logits_process.py (hunk 6 above)
Comment (amyeroberts) [reply src/transformers/generation/logits_process.py:1718]: cheeky :D

**Review by amyeroberts — APPROVED** [review]
Thanks for working on this! 

Happy with the changes - only concern is the difference in the whisper processor @sanchit-gandhi can you confirm this?

File: src/transformers/generation/logits_process.py (hunk 4 above)
Comment (sanchit-gandhi) [reply src/transformers/generation/logits_process.py:1650]: Looks good to me - thanks for the updated @gante!

**Review by sanchit-gandhi — APPROVED** [review]
Thanks for the fixes @gante!
